src/nbl/asset/utils/CSPIRVIntrospector.cpp

```diff
@@ -13,472 +13,528 @@ namespace nbl::asset
+            "UNKNOWN_OR_STRUCT",
+            "U64",
+            "I64",
+            "U32",
+            "I32",
+            "U16",
+            "I16",
+            "U8",
+            "I8",
+            "F64",
+            "F32",
+            "F16"
+        };
+
+        return typeNames[static_cast<uint32_t>(varType)];
     };
-    return convert[_imgfmt];
-}
 }//anonymous ns
 
-core::smart_refctd_ptr<const CSPIRVIntrospector::CIntrospectionData> CSPIRVIntrospector::introspect(const SIntrospectionParams& params, bool insertToCache)
+static CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE spvcrossType2E_TYPE(spirv_cross::SPIRType::BaseType basetype)
 {
-    if (!params.cpuShader)
-        return nullptr;
-    
-    if (params.cpuShader->getContentType()!=IShader::E_CONTENT_TYPE::ECT_SPIRV)
-        return nullptr;
-
-    auto introspectionData = m_introspectionCache.find(params);
-    if (introspectionData != m_introspectionCache.end())
-        return introspectionData->second;
-
-    const ICPUBuffer* spv = params.cpuShader->getContent();
-    spirv_cross::Compiler comp(reinterpret_cast<const uint32_t*>(spv->getPointer()), spv->getSize()/4u);
-    auto introspection = doIntrospection(comp,params.entryPoint,params.cpuShader->getStage());
-    
-    if (insertToCache)
-        m_introspectionCache[params] = introspection;
-
-    return introspection;
+    switch (basetype)
+    {
+    case spirv_cross::SPIRType::Int:
+        return CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE::I32;
+    case spirv_cross::SPIRType::UInt:
+        return CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE::U32;
+    case spirv_cross::SPIRType::Float:
+        return CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE::F32;
+    case spirv_cross::SPIRType::Int64:
+        return CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE::I64;
+    case spirv_cross::SPIRType::UInt64:
+        return CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE::U64;
+    case spirv_cross::SPIRType::Double:
+        return CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE::F64;
+    default:
+        return CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE::UNKNOWN_OR_STRUCT;
+    }
 }
 
-bool CSPIRVIntrospector::introspectAllShaders(core::smart_refctd_ptr<const CIntrospectionData>* outIntrospections, const std::span<const ICPUShader::SSpecInfo> infos)
+// returns true if successfully added all the info to self, false if incompatible with what's already in our pipeline or incomplete (e.g. missing spec constants)
+NBL_API2 bool CSPIRVIntrospector::CPipelineIntrospectionData::merge(const CSPIRVIntrospector::CStageIntrospectionData* stageData, const ICPUShader::SSpecInfoBase::spec_constant_map_t* specConstants)
 {
-    auto it = outIntrospections;
-    for (auto info : infos)
+    // validate if descriptors are compatible
+    DescriptorSetBindings descriptorsToMerge[ICPUPipelineLayout::DESCRIPTOR_SET_COUNT];
```

Comment:
you need to initialize those to a copy of the already present `m_descriptorSetBindings` because you're supposed to check cross-stage-introspection collisions as well, not just self